design/cdk-assets.md

```diff
@@ -0,0 +1,248 @@
+
+* **Identifiers:** assets are identified throughout the system via a unique identifier (the key in the `files` and `images` map). This identifier is based on the sha256 of the source (the contents of the file or directory) and will change only if the source changes. It can be used for local caching and optimization purposes. For example, a zip of a directory can be stored in a local cache by this identifier to avoid duplicate work.
+
+* **Sources:** the `source` information for each asset defines the file or directory (relative to the directory of `assets.json`), and additional **packaging** instructions, such as whether to create a zip file from a directory (for file assets) or which options to pass to `docker build`.
+
+* **Destinations:** describe where the asset should be published. At a minimum, for file assets, it includes the S3 bucket and object key and for docker images it includes the repository and image names. A destination may also indicate that an IAM role must be assumed in order to support cross environment publishing. 
+
+  > Destinations are intentionally denormalized in order to keep the logic of where assets are published at the application or framework level and not in this tool. For example, consider a deployment system which requires that all assets are always published to the same location, and then replicated through some other means to their actual consumption point. Alternatively, a user may have unique security requirements that will require certain assets to be stored in dedicated locations (e.g. with a specific key) and others in a different location, even if they all go to the same environment. Therefore, this tool should not take any assumptions on where assets should be published besides the exact instructions in this file.
+
+Here is the complete manifest file schema in typescript:
+
+```ts
+interface AssetManifest {
+  readonly version: 'assets-1.0';
+  readonly files?: { [id: string]: FileAsset };
+  readonly images?: { [id: string]: ImageAsset };
+}
+
+interface FileAsset {
+  readonly source: FileAssetSource;
+  readonly destinations: FileAssetDestination[];
+}
+
+interface ImageAsset {
+  readonly source: ImageAssetSource;
+  readonly destinations: ImageAssetDestination[];
+}
+
+interface FileAssetSource {
+  readonly file: string;                                // file or directory name, relative to basedir
+  readonly packaging?: FileAssetPackaging;              // packaging (default "FILE")
+}
+
+enum FileAssetPackaging {
+  FILE = 'file',                                        // just upload "file" as-is
+  ZIP_DIRECTORY = 'zip'                                 // zip the directory and then upload
+}
+
+interface FileAssetDestination {
+  readonly assumeRoleArn?: string;                      // iam role to assume
+  readonly assumeRoleExternalId?: string;               // external id to pass to assume-role
```

Comment:
My assumption is either a user will run this locally or a CodeBuild step will be running this publish step. Either way, you shouldn't need an External ID (i.e. there's no deputy to confuse since the role won't be assumed by a service - even CodeBuild's service role is a customer provided role). 

Did you have some other idea on how this would be used?

Reply:
Since the assumed role can be in a different account, I think it's a good practice to allow users to specify the external ID if they wish to specify it in the cross-account trust policy. They don't have to use it of course.